### DISCUSSION TITLE
Integrating recast & detour to project

### DISCUSSION THREAD

**Discussion by TheExeQ** [recastnavigation/recastnavigation#583] [Q&A] [answered]
Hello! Im trying to integrate rc and dt to my project however im a bit stuck trying to figure out how to use it. I have the library itself included in my project but im not sure how to actually use it.
I've read the README where it says "If you are new to Recast & Detour, check out Sample_SoloMesh.cpp to get started with building navmeshes and NavMeshTesterTool.cpp to see how Detour can be used to find paths." 

I have checked out both of them but there is just so much so i don't know what is necessery for it to work. Like i tried copying the Sample_SoloMesh to my project but then it wanted InputGeom and Sample which needed SampleInterfaces etc.. idk how much of this is needed to just get the core to work.

Like i already have a modelloader from which i can get verts, indices, normals etc.. Can i use this info to use recast or do i actually need InputGeom and everything else?

EDIT: I figured it out after a while of experimenting :P

**Comment by grahamboree** [comment, marked as answer]
Hi!  I'm in the process of writing up a more detailed intro guide to Recast, but the gist of it is:

> Like i already have a modelloader from which i can get verts, indices, normals etc.. Can i use this info to use recast or do i actually need InputGeom and everything else?

Yes, you can use your modelloader!  You likely don't need 95% of what's in RecastDemo.

The demo code tries to be a super full featured demo that uses every feature Recast has to offer, and keep intermediate results so you can examine them in the UI, and track build timing information.  There's a lot of code in there you probably don't need.

The relevant code is probably just these lines in `Sample_SoloMesh::handleBuild`: https://github.com/recastnavigation/recastnavigation/blob/d0b2ed8ff5d371f375cf49dad7ca96e8b22142d0/RecastDemo/Source/Sample_SoloMesh.cpp#L388-L641

For just building a navmesh you essentially need to call the following functions:
```
rcCalcGridSize // Figure out how big the raster voxel grid will be

// Create a voxel heightfield
rcAllocHeightfield
rcCreateHeightfield
rcMarkWalkableTriangles
rcRasterizeTriangles

// Clean up the rasterized data and filter out non-walkable areas.
rcFilterLowHangingWalkableObstacles
rcFilterLedgeSpans
rcFilterWalkableLowHeightSpans

// Create a compact heightfield
rcAllocCompactHeightfield
rcBuildCompactHeightfield

// Further refine the voxel representation
rcErodeWalkableArea
rcBuildDistanceField
rcBuildRegions

// Create a set of polygon contours
rcAllocContourSet
rcBuildContours

// Triangulate polygons from the contours
rcAllocPolyMesh
rcBuildPolyMesh

// Build a navmesh from the triangulated polygons
rcAllocPolyMeshDetail
rcBuildPolyMeshDetail

// Cleanup
rcFreeHeightField
rcFreeCompactHeightfield
rcFreeContourSet
```

There's a lot happening here, but that's by design.  It's extremely common to want to tie into the navmesh building process at certain points to add things like spatial annotations, cover points, etc. so Recast airs on the side of being overly verbose, with the goal of being easily customizable.  Long term I'd like to add a higher-level API that's a bit simpler to use, but for now this is about as concise as it gets.

**Reply by alishahb** [reply]
It would be great to have documentation, so we could improve on our settings with better understanding the config!

**Reply by SigmaKlim** [reply]
Hi, I am new to rc&dt. I am trying to integrate it into my project. I have transported code from handleBuild in RecastDemo. Everything there seems alright. Now I am trying to debug draw navmesh. I am inspecting handleRender in RecastDemo and I can't understand much there. Any tips or words of advice maybe? I woder why the devs of recastnavigation did not proide a simpler interface for their API....

**Reply by grahamboree** [reply]
Hi!  The demo classes draw the navmesh through calls to functions that are defined in `DebugUtils`.  The drawing code isn't terribly complex but the demo classes have a lot of options that let you draw the intermediate step data, as well as some setup for drawing the input mesh, etc.  If you want to just draw a navmesh, you can call `duDebugDrawNavMeshPolysWithFlags`.  There are other features like offmesh connections that require similar debug draw calls.
While a simple API is certainly our goal, we're also making a pretty industrial-strength tool here and customization is most important to us.  I agree there should be a layer on top of the API currently provided that's more beginner-friendly, but we also want to provide all the options to those people integrating this into complex engines and games.  It's a tricky balance but [a high-level API is on our roadmap](https://recastnav.com/md_Docs_Roadmap.html).  You're welcome to help out with that if you're passionate about it :)
